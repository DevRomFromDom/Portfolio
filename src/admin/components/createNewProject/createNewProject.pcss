.from-container {
    margin: 19px 0;
}
.form-cols {
    display: flex;
    @media screen and (max-width: 768px) {
        flex-direction: column;
        justify-content: center;
        align-items: center;
    }
}

.form-col {
    width: 50%;
    margin-right: 30px;

    &:last-child {
        margin-right: 0;
    }
    @media screen and (max-width: 768px) {
        width: 100%;
        max-width: 500px;
        margin-right: 0;
        margin-bottom: 30px;
        &:last-child {
            margin-bottom: 0;
        }
    }
}

.uploader {
    min-height: 276px;
    display: flex;
    align-items: center;
    justify-content: center;
    flex-direction: column;
    background: #dee4ed;
    border: 1px dashed #a1a1a1;
    padding: 0 23%;
    text-align: center;
    position: relative;
    background-size: cover;

    &.active {
        .uploader-title {
            background: rgba(#fff, 1);
            overflow: hidden;
            opacity: 1;
            border-radius: 10px;
        }
    }
    &.warn {
        border-color: red;
        background: rgba(255, 0, 0, 0.137);
        .uploader-title {
            color: red;
        }
    }

    &.hovered {
        border-color: blue;
    }
}

.uploader-title {
    color: $text-color;
    opacity: 0.5;
    margin-bottom: 30px;
    font-size: 16px;
    line-height: 20px;
}

.form-row {
    margin-bottom: 30px;
}

.form-buttons {
    display: flex;
    justify-content: flex-end;
    width: 100%;
    @media screen and (max-width: 768px) {
        justify-content: center;
    }
}

.btn {
    margin-right: 10px;

    &:last-child {
        margin-right: 0;
    }
}
.create-project-component {
    width: 100%;
    padding-right: 28px;
    margin-bottom: 28px;
    @media screen and (max-width: 768px) {
      padding-right: 0;
    }
    @media screen and (max-width: 480px) {
        padding: 0;

    }
}
.edit-button {
    padding: 15px 0;
    text-align: center;
    font-size: 16px;
    line-height: 22px;
    color: #383bcf;
}
