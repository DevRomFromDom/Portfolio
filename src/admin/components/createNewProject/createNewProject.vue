<template>
    <div class="create-project-component">
        <card :title="modalTitle">
            <form
                slot="content"
                class="from-container"
                @submit.prevent="handleSubmit"
            >
                <div class="form-cols">
                    <div class="form-col">
                        <div class="image-load">
                            <label
                                :style="{
                                    backgroundImage: `url(${bgImage})`,
                                }"
                                :class="[
                                    'uploader',
                                    { active: preview },
                                    { hovered: hovered },
                                    { warn: noPhoto },
                                ]"
                                @dragover="handleDragOver"
                                @dragleave="hovered = false"
                                @drop="handleChange"
                            >
                                <div class="uploader-title" v-if="!editData.id">
                                    Перетащите или загрузите для загрузки
                                    изображения
                                </div>
                                <div class="uploader-btn" v-if="!editData.id">
                                    <app-button
                                        typeAttr="file"
                                        @change="handleChange"
                                        :title="
                                            preview === ''
                                                ? 'ЗАГРУЗИТЬ'
                                                : 'ИЗМЕНИТЬ'
                                        "
                                    />
                                </div>
                            </label>
                        </div>
                        <app-button
                            plain
                            class="edit-button"
                            v-if="editData.id"
                            typeAttr="file"
                            @change="handleChange"
                        />
                    </div>
                    <div class="form-col">
                        <div class="form-row">
                            <app-input
                                title="Название"
                                v-model="project.title"
                                :errorMessage="
                                    validation.firstError('project.title')
                                "
                                @input="validation.reset()"
                            />
                        </div>
                        <div class="form-row">
                            <app-input
                                title="Ссылка"
                                v-model="project.link"
                                :errorMessage="
                                    validation.firstError('project.link')
                                "
                                @input="validation.reset()"
                            />
                        </div>
                        <div class="form-row">
                            <app-input
                                field-type="textarea"
                                title="Описание"
                                v-model="project.description"
                                :errorMessage="
                                    validation.firstError('project.description')
                                "
                                @input="validation.reset()"
                            />
                        </div>
                        <div class="form-row">
                            <tags-adder
                                v-model="project.techs"
                                :valError="
                                    validation.firstError('project.techs')
                                "
                                @input="validation.reset()"
                            />
                        </div>
                    </div>
                </div>
                <div class="form-buttons">
                    <div class="btn" @click="resetChanges">
                        <app-button title="Отмена" plain />
                    </div>
                    <div class="btn">
                        <app-button title="Сохранить" />
                    </div>
                </div>
            </form>
        </card>
    </div>
</template>

<script>
import card from "../card/card.vue";
import appButton from "../button";
import TagsAdder from "../tagsAdder/tagsAdder.vue";
import appInput from "../input";
import { mapActions } from "vuex";
import { Validator, mixin as ValidatorMixin } from "simple-vue-validator";
export default {
    mixins: [ValidatorMixin],
    validators: {
        "project.title": (value) => {
            return Validator.value(value).required("Введите название работы");
        },
        "project.link": (value) => {
            return Validator.value(value).required("Укажите ссылку на работу");
        },
        "project.description": (value) => {
            return Validator.value(value).required("Введите описание работы");
        },
        "project.techs": (value) => {
            return Validator.value(value).required("Введите тэги");
        },
    },
    components: { card, appButton, TagsAdder, appInput },
    props: {
        modalTitle: String,
        editData: Object,
    },
    created() {
        if (this.editData.id !== undefined) {
            this.project = { ...this.editData };
        } else {
            this.project = {
                title: "",
                link: "",
                description: "",
                techs: "",
                photo: {},
            };
        }
    },
    data() {
        return {
            hovered: false,
            project: {},
            preview: "",
            noPhoto: "",
        };
    },
    methods: {
        ...mapActions({
            addNewProjectAction: "projects/addNewProject",
            editProjectAction: "projects/editProject",
        }),
        resetChanges(e) {
            e.preventDefault();
            this.$emit("reset");
        },
        handleDragOver(e) {
            e.preventDefault();
            this.hovered = true;
        },
        async handleSubmit(e) {
            e.preventDefault();

            if (
                (await this.$validate()) === false ||
                this.preview.length === 0
            ) {
                this.preview.length === 0
                    ? (this.noPhoto = true)
                    : (this.noPhoto = "");
                return;
            }

            if (this.project.id === undefined) {
                this.addNewProjectAction({
                    title: this.project.title,
                    techs: this.project.techs,
                    photo: this.project.photo,
                    link: this.project.link,
                    description: this.project.description,
                });
            } else {
                this.editProjectAction(this.project);
            }
            this.$emit("closeModal");
        },
        handleChange(e) {
            e.preventDefault();
            const file = e.dataTransfer
                ? e.dataTransfer.files[0]
                : e.target.files[0];

            this.project.photo = file;

            this.renderPhoto(file);
            this.hovered = false;
        },
        renderPhoto(file) {
            const reader = new FileReader();
            reader.readAsDataURL(file);
            reader.onloadend = () => {
                this.noPhoto = false;
                this.preview = reader.result;
            };
        },
    },
    computed: {
        bgImage() {
            if (typeof this.project.photo === "string") {
                this.preview = this.project.photo;
                return `https://webdev-api.loftschool.com/${this.project.photo}`;
            } else {
                return `${this.preview}`;
            }
        },
    },
};
</script>

<style lang="postcss" scoped src="./createNewProject.pcss"></style>
