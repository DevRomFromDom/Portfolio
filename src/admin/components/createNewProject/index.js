export {default} from "./createNewProject.vue"