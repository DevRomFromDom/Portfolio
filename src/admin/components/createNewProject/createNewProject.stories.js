import createNewProject from "./createNewProject.vue";

export default {
    title: "createNewProject",
    components: { createNewProject },
};

export const defaultView = () => ({
    components: { createNewProject },
    props: {
        editData: {
            id: 3434,
            title: "title",
            link: "link",
            techs: "fgfg, fgfg, dfgffg",
            photo: {},
        },
    },
    template: `
        <createNewProject :editData="editData">
        </createNewProject>
    `,
});
