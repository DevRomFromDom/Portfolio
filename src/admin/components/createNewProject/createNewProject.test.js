import createNewProject from "./createNewProject.vue";
import { mount } from "@vue/test-utils";


it("CreateNewProject has been mounted", async () => {
    const wrapper = mount(createNewProject, {propsData: {editData: {
        title: "Title",
        link: "link",
        description: "dec",
        techs: "fgf, fgfg, fgfg",
        photo: {},
        id: 123,
    }}});
    expect(wrapper.element).toMatchSnapshot();
});
