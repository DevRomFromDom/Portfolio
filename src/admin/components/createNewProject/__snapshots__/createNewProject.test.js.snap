// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateNewProject has been mounted 1`] = `
<div
  class="create-project-component"
>
  <div
    class="card-component"
  >
    <div
      class="header"
    >
      <div
        class="text"
      />
       
    </div>
     
    <div
      class="content"
    >
      <form
        class="from-container"
      >
        <div
          class="form-cols"
        >
          <div
            class="form-col"
          >
            <div
              class="image-load"
            >
              <label
                class="uploader"
                style="background-image: url();"
              >
                <!---->
                 
                <!---->
              </label>
            </div>
             
            <!---->
          </div>
           
          <div
            class="form-col"
          >
            <div
              class="form-row"
            >
              <label
                class="input input_labeled"
              >
                <div
                  class="title"
                >
                  Название
                </div>
                 
                <input
                  class="input__elem field__elem"
                />
                 
                <div
                  class="input__error-tooltip"
                >
                  <div
                    class="input__error-tooltip-container"
                  >
                    
                  </div>
                </div>
              </label>
            </div>
             
            <div
              class="form-row"
            >
              <label
                class="input input_labeled"
              >
                <div
                  class="title"
                >
                  Ссылка
                </div>
                 
                <input
                  class="input__elem field__elem"
                />
                 
                <div
                  class="input__error-tooltip"
                >
                  <div
                    class="input__error-tooltip-container"
                  >
                    
                  </div>
                </div>
              </label>
            </div>
             
            <div
              class="form-row"
            >
              <label
                class="textarea"
              >
                <div
                  class="title"
                >
                  Описание
                </div>
                 
                <textarea
                  class="textarea__elem field__elem"
                />
                 
                <div
                  class="input__error-tooltip"
                >
                  <div
                    class="input__error-tooltip-container"
                  >
                    
                  </div>
                </div>
              </label>
            </div>
             
            <div
              class="form-row"
            >
              <div
                class="tags-adder-component"
              >
                <label
                  class="input input_labeled"
                >
                  <div
                    class="title"
                  >
                    Добавление тэга
                  </div>
                   
                  <input
                    class="input__elem field__elem"
                  />
                   
                  <div
                    class="input__error-tooltip"
                  >
                    <div
                      class="input__error-tooltip-container"
                    >
                      
                    </div>
                  </div>
                </label>
                 
                <ul
                  class="tags"
                >
                  <li
                    class="tag"
                  >
                    <div
                      class="tag interactive"
                    >
                      <span>
                        fgf
                      </span>
                       
                      <button
                        class="remove"
                        type="button"
                      />
                    </div>
                  </li>
                  <li
                    class="tag"
                  >
                    <div
                      class="tag interactive"
                    >
                      <span>
                        fgfg
                      </span>
                       
                      <button
                        class="remove"
                        type="button"
                      />
                    </div>
                  </li>
                  <li
                    class="tag"
                  >
                    <div
                      class="tag interactive"
                    >
                      <span>
                        fgfg
                      </span>
                       
                      <button
                        class="remove"
                        type="button"
                      />
                    </div>
                  </li>
                </ul>
              </div>
            </div>
          </div>
        </div>
         
        <div
          class="form-buttons"
        >
          <div
            class="btn"
          >
            <!---->
          </div>
           
          <div
            class="btn"
          >
            <!---->
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
